### PR TITLE
Generate receivers from alertmanagerconfig even if routes are empty

### PR DESCRIPTION
**Pull request by abhilashjoseph**
This PR is to fix the issue logged here https://github.com/prometheus-operator/prometheus-operator/issues/3819. The Alertmanager config generation ignores the receiver configuration when no routes are present in the AlertManagerConfig definition. This changes the order of generation of routes (to avoid the loop skip, in case route is absent)

**Release Note Template (will be copied)**

```release-note:
[BUGFIX] Generate receivers from alertmanagerconfig even if routes are empty. #3819
```

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,7 @@
+## 0.45.1 / 2021-02-01
+
+* [BUGFIX] Generate receivers from alertmanagerconfig even if routes are empty. #3819
```
Comment (s-urbaniak) [inline CHANGELOG.md:3]: nit: this is not necessary as we generate the changelog from the PR description.

File: VERSION [hunk 2]
```diff
@@ -1 +1 @@
-0.45.0
+0.45.1
```
Comment (s-urbaniak) [inline VERSION:1]: see above, the version will be adjusted once we cut a release.

File: pkg/alertmanager/amcfg.go [hunk 3]
```diff
@@ -96,20 +96,20 @@ func (cg *configGenerator) generateConfig(
...
 		for _, receiver := range amConfigs[amConfigIdentifier].Spec.Receivers {
 			receivers, err := cg.convertReceiver(ctx, &receiver, crKey)
 			if err != nil {
 				return nil, errors.Wrapf(err, "AlertmanagerConfig %s", crKey.String())
 			}
 			baseConfig.Receivers = append(baseConfig.Receivers, receivers)
 		}
+
+		// Skip early if there's no route definition.
+		if amConfigs[amConfigIdentifier].Spec.Route == nil {
```
Comment (s-urbaniak) [inline pkg/alertmanager/amcfg.go:108]: nit: maybe it is clearer if we simply do:
```
if amConfigs[amConfigIdentifier].Spec.Route != nil {
  subRoutes = append(subRoutes, convertRoute(amConfigs[amConfigIdentifier].Spec.Route, crKey, true))
{
```

**Review by s-urbaniak — CHANGES_REQUESTED** [review]
just a few nits and cc @simonpasquier to have another pair of eyes.

**Comment by s-urbaniak** [comment]
@abhilashjoseph thank you for the contribution! :tada: I just modified the PR description to include the change.

**Comment by s-urbaniak** [comment]
another review comment: do you mind to add some unit tests in `amcfg_test.go` as well? thank you!

**Review by simonpasquier — COMMENTED** [review]
IIUC you want to declare an `AlertmanagerConfig` resource like this:

```yaml
apiVersion: monitoring.coreos.com/v1alpha1
kind: AlertmanagerConfig
metadata:
  name: example
  namespace: default
spec:
  receivers:
  - name: 'webhook'
    webhookConfigs:
    - url: http://example.com/
```
And then reference it in the global Alertmanager config secret:

```
route:
  group_by: ['job']
  group_wait: 30s
  group_interval: 5m
  repeat_interval: 12h
  receiver: default-example-webhook
```

Is this correct? IMO it would be a bit fragile and difficult to maintain because it relies on how the operator generates receiver names. Maybe you're interested by #3737 which aims at leveraging `AlertmanagerConfig` for the global Alertmanager configuration.

File: CHANGELOG.md (hunk 1 above)
Comment (abhilashjoseph) [reply CHANGELOG.md:3]: Thank you @s-urbaniak , will remove this

**Comment by abhilashjoseph** [comment]
> IIUC you want to declare an `AlertmanagerConfig` resource like this:
> 
> ```yaml
> apiVersion: monitoring.coreos.com/v1alpha1
> kind: AlertmanagerConfig
> metadata:
>   name: example
>   namespace: default
> spec:
>   receivers:
>   - name: 'webhook'
>     webhookConfigs:
>     - url: http://example.com/
> ```
> 
> And then reference it in the global Alertmanager config secret:
> 
> ```
> route:
>   group_by: ['job']
>   group_wait: 30s
>   group_interval: 5m
>   repeat_interval: 12h
>   receiver: default-example-webhook
> ```
> 
> Is this correct? IMO it would be a bit fragile and difficult to maintain because it relies on how the operator generates receiver names. Maybe you're interested by #3737 which aims at leveraging `AlertmanagerConfig` for the global Alertmanager configuration.

Thanks @simonpasquier, yes that issue is exactly why I ended up in making this change. Currently the mandatory namespace match at first level route, is forcing us to create the alertmanager config at all namespaces and also some alerts don't have the namespace label. The change to get the namespace label configurable would also be more involved because, it needs to deal with duplicate alert manager configs and overlap of alerts in the case where you have global alertmanagerconfig vs alertmanagerconfigs in each ns

File: VERSION (hunk 2 above)
Comment (abhilashjoseph) [reply VERSION:1]: @s-urbaniak should the go.mod change to the patch version be rolled back as well?

**Comment by abhilashjoseph** [comment]
> another review comment: do you mind to add some unit tests in `amcfg_test.go` as well? thank you!

Added unit test!

**Comment by abhilashjoseph** [comment]
@simonpasquier, @s-urbaniak, please let me know if I should keep this request open, This could be a very short term unblocker for the issue that many of us are facing, while we work on the global alertmanager config changes?

**Comment by PhilipGough** [comment]
@abhilashjoseph - can this be closed now since https://github.com/prometheus-operator/prometheus-operator/commit/d41eb193e5bb250e1a1b0c60a120e4e78ab59765 is merged?

**Comment by simonpasquier** [comment]
closed by #4220

**Pull request closed without merging**